types/redux-promise-middleware/index.d.ts

```diff
@@ -2,7 +2,23 @@
 // Project: https://github.com/pburtchaell/redux-promise-middleware
 // Definitions by: ianks <https://github.com/ianks>
 // Definitions: https://github.com/borisyankov/DefinitelyTyped
+// TypeScript Version: 2.8
 
 import { Middleware } from 'redux';
 
 export default function promiseMiddleware(config?: { promiseTypeSuffixes?: string[], promiseTypeDelimiter?: string }): Middleware;
+
+declare module "redux" {
+  export interface PromiseAction<R> extends Action {
```

Comment:
Does redux-promise-middleware make this change to `Dispatch`? Or should this change go into redux' types?

Reply:
Hello @sandersn,

From my understanding of these libraries, the promise middleware will return [a promise object](https://github.com/pburtchaell/redux-promise-middleware/blob/master/src/index.js#L203) whereas redux will [return the action](https://github.com/reduxjs/redux/blob/master/src/createStore.js#L197) instead. So yes, I believe that this change should be here, as an override of `Dispatch`. I was inspired from what was done in [redux-thunk 2.2.0](https://github.com/reduxjs/redux-thunk/blob/v2.2.0/index.d.ts#L8) (another middleware for Redux, which makes dispatch accept a function as well as an action).

However, I have just noticed that the last version (2.3.0) of redux-thunk has [changed their type definition](https://github.com/reduxjs/redux-thunk/compare/v2.2.0...v2.3.0#diff-b52768974e6bc0faccb7d4b75b162c99), but I am not sure that understand how it works after their changes.

Reply:
I don’t know anything about redux, but from redux-thunk’s tests, it looks like redux has a new function called applyMiddleware that lets you avoid monkeypatching redux objects. I’m not sure if redux-promise-middleware uses this function yet.

Reply:
If it doesn’t, then your correct change is probably correct.

Reply:
This is not really about monkeypatching redux objects.
In the redux world, a middleware is a function that returns a `Dispatch` function. These functions are chained together, so each middleware's `Dispatch` receives the state of the previous one, and pass it to the next, until redux's dispatch is called.
Usually, a middleware will just return the same value as its underlying function.
However, redux-promise-middleware is returning a promise object instead, and this PR is trying to reflect this.

Reply:
If you're not mutating the state, imitating the new version of redux-thunk is probably the right answer. It looks like `Middleware` has a bunch of type parameters that allow you to specify your custom Dispatch function that way.